avocado_vt/plugins/vt_joblock.py

```diff
@@ -0,0 +1,113 @@
+
+from ..test import VirtTest
+
+
+class VTJobLock(JobPre, JobPost):
+
+    name = 'vt-joblock'
+    description = 'Avocado-VT Job Lock/Unlock'
+
+    def __init__(self):
+        self.log = logging.getLogger("avocado.app")
+        self.lock_dir = settings.get_value(section="plugins.vtjoblock",
+                                           key="dir",
+                                           key_type=str,
+                                           default='/tmp')
+        self.lock_file = None
+
+    def _abort(self, message, job):
+        """
+        Aborts the Job by exiting Avocado, adding a failure to the job status
+        """
+        self.log.error(message)
+        sys.exit(exit_codes.AVOCADO_JOB_FAIL | job.exitcode)
+
+    def _set_lock(self, job):
+        if not os.path.isdir(self.lock_dir):
+            msg = ('VT Job lock directory "%s" does not exist... '
+                   'exiting...' % self.lock_dir)
+            self._abort(msg, job)
+
+        pattern = 'avocado-vt-joblock-%(jobid)s-%(uid)s-%(random)s.pid'
+        # the job unique id is already random, but, let's add yet another one
+        rand = ''.join([random.choice(string.ascii_lowercase + string.digits)
+                        for i in xrange(8)])
+        path = pattern % {'jobid': job.unique_id,
+                          'uid': os.getuid(),
+                          'random': rand}
+        path = os.path.join(self.lock_dir, path)
+
+        try:
+            with open(path, 'w') as lockfile:
+                lockfile.write("%u" % os.getpid())
+            self.lock_file = path
+        except IOError as e:
+            msg = ('Failed to create VT Job lock file "%s". Exiting...' % path)
+            self._abort(msg, job)
+
+        self.log.info("VT LOCK    : %s", self.lock_file)
+
+    def _get_lock_file_pid(self):
+        """
+        Gets the lock file path and the process ID
+
+        If the lock file can not be located, this returns (None, 0).
+
+        :returns: the path and the (integer) process id
+        :rtype: tuple(str, int)
+        """
+        try:
+            files = os.listdir(self.lock_dir)
+            if not files:
+                return (None, 0)
+        except OSError as e:
+            if e.errno == errno.ENOENT:
+                return (None, 0)
+
+        pattern = re.compile(r'avocado-vt-joblock-[0-9a-f]{40}-[0-9]+'
+                             '-[0-9a-z]{8}\.pid')
+        for lock_file in files:
+            if pattern.match(lock_file):
+                path = os.path.join(self.lock_dir, lock_file)
+                if os.path.isfile(path):
+                    content = int(open(path, 'r').read())
+                    pid = int(content)
+                    if pid > 0:
+                        return (path, pid)
+        return (None, 0)
+
+    def _lock(self, job):
+        _, lock_pid = self._get_lock_file_pid()
+        if lock_pid > 0:
+            msg = ("Avocado-VT job lock acquired by PID %u. "
```

Comment:
Yep, I'll send an updated version with those changes.
On May 30, 2016 12:42 PM, "Lukáš Doktor" <notifications@github.com> wrote:

> In avocado_vt/plugins/vt_joblock.py
> <https://github.com/avocado-framework/avocado-vt/pull/543#discussion_r65087037>
> :
>
> > +        pattern = re.compile(r'avocado-vt-joblock-[0-9a-f]{40}-[0-9]+'
> > +                             '-[0-9a-z]{8}\.pid')
> > +        for lock_file in files:
> > +            if pattern.match(lock_file):
> > +                path = os.path.join(self.lock_dir, lock_file)
> > +                if os.path.isfile(path):
> > +                    content = int(open(path, 'r').read())
> > +                    pid = int(content)
> > +                    if pid > 0:
> > +                        return (path, pid)
> > +        return (None, 0)
> > +
> > +    def _lock(self, job):
> > +        _, lock_pid = self._get_lock_file_pid()
> > +        if lock_pid > 0:
> > +            msg = ("Avocado-VT job lock acquired by PID %u. "
>
> This is the place where it's important to log the lockfile as most people
> would see this message. Actually after playing with this for a while I
> think we could remove the VT LOCK :
> /tmp/avocado-vt-joblock-38b2592547c6a133a281ba02d0eeb12959e706bd-1000-e7kmealb.pid
> line and only log the filelock in case of failure (usually this message).
>
> What do you think?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/avocado-framework/avocado-vt/pull/543/files/223f5509233f77c0a8af0e13f5a5175c37e4b4f3#r65087037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAbSyp73fNzTTZ2LTeZ0Ij2nKq_EwS9mks5qGwVsgaJpZM4IotF7>
> .
>